include/internal/quic_record_rx.h

```diff
@@ -0,0 +1,507 @@
+ */
+int ossl_qrx_provide_secret(OSSL_QRX              *qrx,
+                            uint32_t               enc_level,
+                            uint32_t               suite_id,
+                            EVP_MD                *md,
+                            const unsigned char   *secret,
+                            size_t                 secret_len);
+
+/*
+ * Informs the QRX that it can now discard key material for a given EL. The QRX
+ * will no longer be able to process incoming packets received at that
+ * encryption level. This function is idempotent and succeeds if the EL has
+ * already been discarded.
+ *
+ * Returns 1 on success and 0 on failure.
+ */
+int ossl_qrx_discard_enc_level(OSSL_QRX *qrx, uint32_t enc_level);
+
+/*
+ * Packet Reception
+ * ================
+ */
+
+/* Information about a received packet. */
+typedef struct ossl_qrx_pkt_st {
+    /* Opaque handle to be passed to ossl_qrx_release_pkt. */
+    void               *handle;
+
+    /*
+     * Points to a logical representation of the decoded QUIC packet header. The
+     * data and len fields point to the decrypted QUIC payload (i.e., to a
+     * sequence of zero or more (potentially malformed) frames to be decoded).
+     */
+    QUIC_PKT_HDR       *hdr;
+
+    /*
+     * Address the packet was received from. If this is not available for this
+     * packet, this field is NULL (but this can only occur for manually injected
+     * packets).
+     */
+    const BIO_ADDR     *peer;
+
+    /*
+     * Local address the packet was sent to. If this is not available for this
+     * packet, this field is NULL.
+     */
+    const BIO_ADDR     *local;
+
+    /*
+     * This is the length of the datagram which contained this packet. Note that
+     * the datagram may have contained other packets than this. The intended use
+     * for this is so that the user can enforce minimum datagram sizes (e.g. for
+     * datagrams containing INITIAL packets), as required by RFC 9000.
+     */
+    size_t              datagram_len;
+
+    /* The PN which was decoded for the packet, if the packet has a PN field. */
+    QUIC_PN             pn;
+
+    /*
+     * Time the packet was received, or ossl_time_zero() if the demuxer is not
+     * using a now() function.
+     */
+    OSSL_TIME           time;
+} OSSL_QRX_PKT;
+
+/*
+ * Tries to read a new decrypted packet from the QRX.
+ *
+ * On success, all fields of *pkt are filled and 1 is returned.
+ * Else, returns 0.
+ *
+ * The resources referenced by pkt->hdr, pkt->hdr->data and pkt->peer will
+ * remain allocated at least until the user frees them by calling
+ * ossl_qrx_release_pkt, which must be called once you are done with the packet.
+ */
+int ossl_qrx_read_pkt(OSSL_QRX *qrx, OSSL_QRX_PKT *pkt);
+
+/*
+ * Release the resources pointed to by an OSSL_QRX_PKT returned by
+ * ossl_qrx_read_pkt. Pass the opaque value pkt->handle returned in the
+ * structure.
+ */
+void ossl_qrx_release_pkt(OSSL_QRX *qrx, void *handle);
```

Comment:
Is there a reason why the opaque handle must be passed?  Can't we just as well pass the `OSSL_QRX_PKT` pointer and leave it to `ossl_qrx_release_pkt()` to use the proper bits?

Reply:
We could. I guess the issue there is that it requires the caller to keep `OSSL_QRX_PKT` around until it wants to release the resources. That isn't really the intention, since `OSSL_QRX_PKT` is intended as just a temporary structure for returning information. So you could throw away the structure immediately after calling `ossl_qrx_read_pkt` and store the information in it in another way, so long as you keep the handle.

Reply:
Ok, there seems to be different ideas about this.  See the discussion starting here: https://github.com/openssl/openssl/pull/18838#discussion_r940383537

Reply:
Responded over there